tools/RAiDER/downloadGNSSDelays.py

```diff
@@ -0,0 +1,330 @@
+    Inputs: 
+        bbox    - length-4 list of floats that describes a bounding box. Format is 
+                  S N W E
+    '''
+    if writeLoc is None:
+        writeLoc = os.path.join(os.getcwd(), 'gnssStationList_overbbox.csv')
+    else:
+        writeLoc = os.path.join(writeLoc, 'gnssStationList_overbbox.csv')
+
+    if userstatList:
+        with open(userstatList) as f:
+            userstatList = [line.rstrip('\n') for line in f]
+
+    statList = getStatsByllh(llhBox = bbox, userstatList = userstatList)
+
+    # write to file and pass final stations list
+    statList.to_csv(writeLoc, index=False)
+    stations = list(statList['ID'].values)
+
+    return stations, writeLoc
+
+
+def inBox(lat, lon, llhbox):
+    '''
+    Checks whether the given lat, lon pair are inside the bounding box llhbox
+    '''
+    lattest = lat < llhbox[1] and lat > llhbox[0]    
+    lontest = lon < llhbox[3] and lon > llhbox[2]
+    if lattest and lontest:
+        return True
+    else:
+        return False
+
+
+def fix_lons(lon):
+    """
+    Fix the given longitudes into the range ``[-180, 180]``.
+
+    """
+    fixed_lon = ((lon + 180) % 360) - 180
+    # Make the positive 180s positive again.
+    if fixed_lon == -180 and lon > 0:
+        fixed_lon *= -1
+
+    return fixed_lon
+
+
+def getID(line):
+    '''
+    Pulls the station ID, lat, and lon for a given entry in the UNR text file
+    '''
+    statID = str(line.split()[0].decode('utf-8'))
+    lat = float(line.split()[1].decode('utf-8'))
+    lon = float(line.split()[2].decode('utf-8'))
+
+    return statID, lat, lon
+
+
+def downloadTropoDelays(stats, years, writeDir = '.', download = False, verbose = False):
+    '''
+    Check for and download GNSS tropospheric delays from an archive. If download is True then 
+    files will be physically downloaded, which again is not necessary as data can be virtually accessed. 
+    '''
+
+    # argument checking
+    if not isinstance(stats, list):
+        if not isinstance(stats, str):
+            raise TypeError('stats should be a string or a list of strings')
+        stats = [stats]
+    if not isinstance(years, list):
+        if not isinstance(years, int):
+            try:
+                years = list(int(years))
+            except TypeError:
+                raise TypeError('years should be an int or a list of ints')
+        years = [years]
+
+    # Iterate over stations and years and check or download data
+    results = []
+    stat_year_tup = itertools.product(stats, years)
+    for s, y in stat_year_tup:
+        if verbose:
+            print('Currently checking station {} in {}'.format(s, y))
+        fileurl = download_UNR(s, y, writeDir = writeDir, download = download, verbose = verbose)
+        # only record valid path
+        if fileurl:
+            results.append({'ID': s, 'year': y, 'path': fileurl})
+    return pd.DataFrame(results).set_index('ID')
+            
+
+def parse_years(input):
```

Comment:
Not exactly sure how the edge cases are supposed to be handled, but I feel like you can make use of the 2 argument version of `range` i.e.
```python3
assert list(range(10, 10)) == []
assert list(range(10, 11)) == [10]
assert list(range(10, 15)) == [10, 11, 12, 13, 14]
assert list(range(20, 10)) == []
```

Reply:
Ok, I was thinking of taking this and leveraging map as so:
```
def parse_years(timestr):
    '''
    Takes string input and returns a list of years as integers
    '''
    years = list(map(int, timestr.split(',')))
    # If two years non-consecutive years are given, query for each year in between
    if len(years) == 2:
         years = list(range(years[0],years[1]+1))
    return years
```

In this way, inputs non-consecutive pairs of years are expanded (e.g. '1994,1997' --> [1994 1995 1996 1997]), while inputs of multiple years are left alone (e.g. '2014,2015,2016' --> [2014 2015 2016])

Reply:
Very nice!